python/dolfinx/fem/assemble.py

```diff
@@ -56,345 +55,175 @@ def _pack(form):
-
-    # Extract (V0, V1) pair for each block in 'a'
-    def fn(form):
-        return form.function_spaces if form is not None else None
-    from functools import partial
-    Vblock = map(partial(map, fn), a)
-
-    # Compute spaces for each row/column block
-    rows = [set() for i in range(len(a))]
-    cols = [set() for i in range(len(a[0]))]
-    for i, Vrow in enumerate(Vblock):
-        for j, V in enumerate(Vrow):
-            if V is not None:
-                rows[i].add(V[0])
-                cols[j].add(V[1])
-
-    rows = [e for row in rows for e in row]
-    cols = [e for col in cols for e in col]
-    assert len(rows) == len(a)
-    assert len(cols) == len(a[0])
-    return rows, cols
-
-
-@ assemble_matrix_block.register(PETSc.Mat)
-def _(A: PETSc.Mat, a: typing.List[typing.List[FormMetaClass]],
-      bcs: typing.List[DirichletBCMetaClass] = [], diagonal: float = 1.0,
-      coeffs=Coefficients(None, None)) -> PETSc.Mat:
-    """Assemble bilinear forms into matrix"""
-    c = (coeffs[0] if coeffs[0] is not None else pack_constants(a),
-         coeffs[1] if coeffs[1] is not None else pack_coefficients(a))
-
-    V = _extract_function_spaces(a)
-    is_rows = _cpp.la.petsc.create_index_sets([(Vsub.dofmap.index_map, Vsub.dofmap.index_map_bs) for Vsub in V[0]])
-    is_cols = _cpp.la.petsc.create_index_sets([(Vsub.dofmap.index_map, Vsub.dofmap.index_map_bs) for Vsub in V[1]])
-
-    # Assemble form
-    for i, a_row in enumerate(a):
-        for j, a_sub in enumerate(a_row):
-            if a_sub is not None:
-                Asub = A.getLocalSubMatrix(is_rows[i], is_cols[j])
-                _cpp.fem.petsc.assemble_matrix(Asub, a_sub, c[0][i][j], c[1][i][j], bcs, True)
-                A.restoreLocalSubMatrix(is_rows[i], is_cols[j], Asub)
-
-    # Flush to enable switch from add to set in the matrix
-    A.assemble(PETSc.Mat.AssemblyType.FLUSH)
-
-    # Set diagonal
-    for i, a_row in enumerate(a):
-        for j, a_sub in enumerate(a_row):
-            if a_sub is not None:
-                Asub = A.getLocalSubMatrix(is_rows[i], is_cols[j])
-                if a_sub.function_spaces[0].id == a_sub.function_spaces[1].id:
-                    _cpp.fem.petsc.insert_diagonal(Asub, a_sub.function_spaces[0], bcs, diagonal)
-                A.restoreLocalSubMatrix(is_rows[i], is_cols[j], Asub)
+    constants = constants or _pack_constants(a)
+    coeffs = coeffs or _pack_coefficients(a)
+    _cpp.fem.assemble_matrix(A, a, constants, coeffs, bcs)
 
+    # If matrix is a 'diagonal'block, set diagonal entry for constrained
+    # dofs
+    if a.function_spaces[0].id == a.function_spaces[1].id:
+        if len(bcs) > 0:
+            warnings.warn("Setting of matrix bc diagonals not yet implemented.")
+    #     A.assemblyBegin(PETSc.Mat.AssemblyType.FLUSH)
+    #     A.assemblyEnd(PETSc.Mat.AssemblyType.FLUSH)
+    #     _cpp.fem.petsc.insert_diagonal(A, a.function_spaces[0], bcs, diagonal)
```

Comment:
Remove commented lines?

Reply:
These are left in as a reminder of what needs doing next.